src/prefect/engine/results/s3_result.py

```diff
@@ -105,89 +105,88 @@ def __getstate__(self) -> dict:
     def __setstate__(self, state: dict) -> None:
         self.__dict__.update(state)
 
-    def write(self) -> str:
+    def write(self, value: Any, **kwargs: Any) -> Result:
         """
         Writes the result to a location in S3 and returns the resulting URI.
 
+        Args:
+            - value (Any): the value to write; will then be stored as the `value` attribute
+                of the returned `Result` instance
+            - **kwargs (optional): if provided, will be used to format the filepath template
+                to determine the location to write to
+
         Returns:
-            - str: the S3 URI
+            - Result: a new Result instance with the appropriately formatted S3 URI
         """
-        if not self._rendered_filepath:
-            raise ValueError("Must call `Result.format()` first")
 
-        self.logger.debug(
-            "Starting to upload result to {}...".format(self._rendered_filepath)
-        )
-        binary_data = self.serialize()
+        new = self.format(**kwargs)
+        new.value = value
+        self.logger.debug("Starting to upload result to {}...".format(new.filepath))
+        binary_data = new.serialize_to_bytes(new.value)
 
         stream = io.BytesIO(binary_data)
 
         ## upload
-        self.client.upload_fileobj(
-            stream, Bucket=self.bucket, Key=self._rendered_filepath
-        )
-        self.logger.debug(
-            "Finished uploading result to {}.".format(self._rendered_filepath)
-        )
+        self.client.upload_fileobj(stream, Bucket=self.bucket, Key=new.filepath)
```

Comment:
Should the upload also handle exceptions similar to https://github.com/PrefectHQ/prefect/blob/3bc254d759b4a86d215b320ce14085898085e1f8/src/prefect/environments/storage/s3.py#L170-L178 where we log the error then raise?

Reply:
Yea I think we should actually - I updated the exception handling logic within both the read and write methods.